### PR TITLE
feat: Implement print plan and basic checks for publish command

### PR DESCRIPTION
**Pull request by popzxc**
This PR adds two features I found useful recently:

- Ability to perform basic checks before running `cargo publish`.
- Ability to just print the packages in order in which they should be published.

The first one proved handy when I had to publish a lot of crates written by different people that weren't initially meant to be published. So every so often a field would be forgotten, but I won't know about it until I'm in the middle of the publishing process.

The second one is useful for large workspaces: crates.io has pretty strict rate limits on publishing new packages, so it's actually easier to get the order in which crates should be published and then do it manually.

### CODE REVIEW

File: cargo-workspaces/src/publish.rs [hunk 1]
```diff
@@ -119,7 +139,19 @@ impl Publish {
                 IndexUrl::crates_io(None, None, None)?
             };
 
-            if is_published(&http_client, index_url, &name, version)? {
+            let already_published = is_published(&http_client, index_url, &name, version)?;
+
+            if self.only_print_plan {
+                let msg = if already_published {
```
Comment (popzxc) [inline cargo-workspaces/src/publish.rs:145]: This makes the printing process longer, but in my use-case (manual workspace publishing) it was handy, as it acts as a reminder of where you stop.

File: cargo-workspaces/src/publish.rs [hunk 2]
```diff
@@ -54,6 +54,17 @@ pub struct Publish {
     /// Don't remove dev-dependencies while publishing
     #[clap(long)]
     no_remove_dev_deps: bool,
+
+    /// Only print the order in which crates should be published.
+    #[clap(long)]
+    only_print_plan: bool,
```
Comment (pksunkara) [inline cargo-workspaces/src/publish.rs:60]: Can you please make this (plan printing) a separate subcommand (`plan`)?

File: cargo-workspaces/src/publish.rs [hunk 3]
```diff
@@ -54,6 +54,17 @@ pub struct Publish {
...
+
+    /// Perform basic checks to make sure that crate can be published,
+    /// e.g. that each package has description and license fields.
+    ///
+    /// This is only a heuristic to prevent simple mistakes.
+    #[clap(long)]
+    basic_checks: bool,
```
Comment (pksunkara) [inline cargo-workspaces/src/publish.rs:67]: Does `cargo publish` not run these checks?

**Review by pksunkara — COMMENTED** [review]
Thanks for the PR! Made a few blocking comments.

File: cargo-workspaces/src/publish.rs (hunk 3 above)
Comment (popzxc) [reply cargo-workspaces/src/publish.rs:67]: It certainly does, but when publishing a workspace, if, say, 7th package has some field missing, you will only know when you're in the middle of process (when you already bumped a version, created a commit, etc). So this is not very useful for publishing individual packages, but IMHO very useful for publishing a workspace.

Comment (pksunkara) [reply cargo-workspaces/src/publish.rs:67]: What do you think of running cargo's `--dry-run` option then instead of writing the checks on our own? That way, we can run our own `--dry-run` for this `publish` subcommand. Here's the related issue https://github.com/pksunkara/cargo-workspaces/issues/4

Comment (popzxc) [reply cargo-workspaces/src/publish.rs:67]: Unfortunately, `--dry-run` doesn't perform these checks, as it aborts before uploading :( 
These checks are performed on the server only.

Comment (pksunkara) [reply cargo-workspaces/src/publish.rs:67]: That makes sense. Let's rename the option to `dry_run`, keep the checks and also incorporate cargo's `--dry-run`. Wouldn't completely fix #4, but is a start towards it.

File: cargo-workspaces/src/publish.rs (hunk 2 above)
Comment (popzxc) [reply cargo-workspaces/src/publish.rs]: Done. Also extracted some common code to utils to prevent code duplication.

File: cargo-workspaces/src/publish.rs (hunk 3 above)
Comment (popzxc) [reply cargo-workspaces/src/publish.rs:52]: Unfortunately, I don't think it would work -- `cargo publish --dry-run` _does_ check that dependencies are published, so `cargo publish --dry-run` would fail for the 2nd crate in the workspace.

And without this functionality, changing the name would probably be misleading.

Comment (pksunkara) [reply cargo-workspaces/src/publish.rs:52]: > And without this functionality, changing the name would probably be misleading.

If we are to incorporate everything else from the cargo's dry run other than checking published dependencies, it wouldn't be misleading, right? I know that cargo-release has a dry run option.

I still think we name this option as such and work toward completing it. But, in the meanwhile, we can document it as in progress.

Comment (popzxc) [reply cargo-workspaces/src/publish.rs]: @pksunkara that makes sense! I've reworked the implementation:

- Renamed `basic-checks` to `dry-run`.
- Changed `dry-run` behavior to ignore versioning and publishing.
- Added `cargo build` as a step together with `basic_checks` so that we check if package compiles at least. This is not strictly equivalent to `--dry-run` behavior, because AFAIU it will reuse the workspace `Cargo.lock`, while with `--dry-run` a fresh `Cargo.lock` is generated for the package. Also, it obviously doesn't try to use previously published packages, but I guess it's unavoidable.
- With `--dry-run` mode for publish, `--basic-checks` do not make much sense in `plan` subcommand, so I removed it there.
- I also made checking if package is published optional for `plan` subcommand, because the difference in speed is significant and you don't always want to check that.

tl;dr:
- `cargo ws publish --dry-run` seems to have a good enough implementation.
- `cargo ws plan` doesn't try to be too smart and only provides information on publishing plan.

File: cargo-workspaces/src/plan.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+pub struct Plan {
+    /// The token to use for accessing the registry
+    #[clap(long, forbid_empty_values(true))]
+    token: Option<String>,
+
+    /// The Cargo registry to check against
+    #[clap(long, forbid_empty_values(true))]
+    registry: Option<String>,
+
+    /// Check if the crates are already published and include the information
```
Comment (pksunkara) [inline cargo-workspaces/src/plan.rs:22]: No need for `and include the information in the output`

File: README.md [hunk 5]
```diff
@@ -267,6 +269,32 @@ OPTIONS:
         --ignore <pattern>    Ignore the crates matched by glob
 ```
 
+### Plan
```
Comment (pksunkara) [inline README.md:272]: Need to update this section

File: README.md [hunk 6]
```diff
@@ -247,6 +248,7 @@ PUBLISH OPTIONS:
         --publish-as-is          Publish crates from the current commit without versioning
         --registry <REGISTRY>    The Cargo registry to use for publishing
         --token <TOKEN>          The token to use for publishing
+        --dry-run                Perform checks without uploading. This option is WIP and performs less checks than `cargo publish --dry-run`.
```
Comment (pksunkara) [inline README.md:251]:
```suggestion
        --dry-run                Perform checks without uploading. WIP and performs fewer checks than `cargo publish --dry-run`.
```

File: cargo-workspaces/src/plan.rs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    /// in the output.
+    #[clap(long)]
```
Comment (pksunkara) [inline cargo-workspaces/src/plan.rs:24]: Add `short`?

File: cargo-workspaces/src/publish.rs [hunk 8]
```diff
@@ -119,7 +125,9 @@ impl Publish {
                 IndexUrl::crates_io(None, None, None)?
             };
 
-            if is_published(&http_client, index_url, &name, version)? {
+            let already_published = is_published(&http_client, index_url, &name, version)?;
+
+            if already_published {
```
Comment (pksunkara) [inline cargo-workspaces/src/publish.rs:130]: I think this chunk can be reverted.

File: cargo-workspaces/src/utils/mod.rs [hunk 9]
```diff
@@ -7,8 +8,10 @@ mod error;
 mod git;
 mod listable;
 mod pkg;
+pub(crate) mod publish_helpers;
```
Comment (pksunkara) [inline cargo-workspaces/src/utils/mod.rs:11]:
```suggestion
mod publish_helpers;
```

and use `pub use` below.

File: cargo-workspaces/src/utils/publish_helpers.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
+//! Helper functions useful when publishing (or preparing for publishing) crates.
```
Comment (pksunkara) [inline cargo-workspaces/src/utils/publish_helpers.rs:1]: File can be named `publish.rs`

File: cargo-workspaces/src/utils/publish_helpers.rs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        .into_iter()
+        .filter(|x| {
+            if let Some((pkg, _)) = pkgs.iter().find(|(p, _)| p.manifest_path == *x) {
+                return pkg.publish.is_none()
+                    || !pkg.publish.as_ref().expect(INTERNAL_ERR).is_empty();
+            }
+
+            false
+        })
+        .collect::<Set<_>>()
```
Comment (pksunkara) [inline cargo-workspaces/src/utils/publish_helpers.rs:34]:
```suggestion
        .collect()
```

File: cargo-workspaces/src/utils/publish_helpers.rs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    external::{
+        http::{HeaderMap, HeaderValue},
+        reqwest::{blocking::Client, header::AUTHORIZATION, Certificate},
+    },
+    index::{ComboIndex, ComboIndexCache, RemoteGitIndex, RemoteSparseIndex},
+    utils::flock::LockOptions,
+    IndexLocation, IndexUrl, KrateName,
+};
+
+pub(crate) fn filter_private(
```
Comment (pksunkara) [inline cargo-workspaces/src/utils/publish_helpers.rs:20]:
```suggestion
pub fn filter_private(
```

We don't need to do `pub(crate)` in this file. We can simply do `pub`

**Review by pksunkara — COMMENTED** [review]
Some minor comments.

The only other thing is I would need some tests for plan (no publishing check) and dry run (without build check).

If you can test the other parts too, great. If not, no worries since they are hard to test.

File: README.md (hunk 6 above)
Comment (pksunkara) [reply README.md:251]: Sorry, wrong place to comment. This needs to be fixed in the comment for this option definition.

**Comment by popzxc** [comment]
@pksunkara I've implemented the requested changes.
However, unfortunately, I cannot run tests, they fail even on `main` branch:

```
failures:
    test_create_bin_2015
    test_create_bin_2018
    test_create_lib_2015
    test_create_lib_2018
    test_member_glob
```

`create_bin` ones seem to fail because `# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html` is missing in the output. However even if I update snapshots there, they panic because of some other error.

Do you have any advice on how to proceed here?

File: README.md [hunk 13]
```diff
@@ -267,6 +269,25 @@ OPTIONS:
...
+```
+USAGE:
+    cargo workspaces plan [OPTIONS]
+
+OPTIONS:
+    -h, --help
+            Print help information
+
+PUBLISH PLAN OPTIONS:
+    --check-published        Check if the crates are already published and include the information in the output
```
Comment (pksunkara) [inline README.md:286]: @popzxc I am working on getting this merged. I think this option would be nice as `--skip-published` (which skips the printing of already published crates). Do you have any issues if I change it like that?

Comment (popzxc) [reply README.md]: Sure

File: cargo-workspaces/src/publish.rs [hunk 14]
```diff
@@ -85,60 +85,52 @@ impl Publish {
...
 
             if is_published(&http_client, index_url, &name, version)? {
                 info!("already published", name_ver);
                 continue;
             }
 
-            if self.no_verify {
+            if self.dry_run {
+                args.push("--dry-run");
+            }
```
Comment (popzxc) [inline cargo-workspaces/src/publish.rs:118]: Do you think it will work? When I checked before, `cargo publish --dry-run` would fail on the 2nd package, because we didn't publish the previous one. This is why I didn't supply this argument before ([as mentioned here](https://github.com/pksunkara/cargo-workspaces/pull/154#discussion_r1672015096)).

Probably this should be removed.

Comment (popzxc) [reply cargo-workspaces/src/publish.rs:118]: Just checked, the command currently doesn't work. See an example output when trying `--dry-run` with `fixtures/normal`:

```
error: failed to prepare local package for uploading

Caused by:
  no matching package named `dep1` found
  location searched: registry `crates-io`
  required by package `top v0.1.0 (/home/popzxc/workspace/cargo-workspaces/fixtures/normal/top)`
```

I will prepare a fix and add tests.

Comment (pksunkara) [reply cargo-workspaces/src/publish.rs:118]: The `--no-verify` was added when `dry-run` is enabled and that would make sure that dependencies are not checked for being published. This is how the other tools in this space are doing it too.

It was working for me in Cargo 1.75. I had a monorepo with half-published crates and all of it worked.

Comment (popzxc) [reply cargo-workspaces/src/publish.rs:118]: Oh, sorry, you're absolutely right! I forgot to reinstall the binary when checking 🤦 
Will revert corresponding changes in #157
